[PyTorch] Refactor Dispatcher to inline less code in fast path

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51368 [PyTorch] Use real if constexpr behind macro in hot template
* #51367 [PyTorch] Debug-gate static_assert in KernelFunction::makeFromUnboxedFunctor
* #51319 [PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor
* #51463 [PyTorch] Fix return value of IValue::to for Tensor/String
* #51315 [PyTorch][easy] Fix TODOs in CppFunction constructors
* #51313 [PyTorch] Reduce template expansion in call_functor_with_args_from_stack
* #51307 [PyTorch] Remove always-empty string args to inferFunctionSchemaFromFunctor
* #51293 [PyTorch] IWYU in torch/csrc/utils/future.h
* #51269 [PyTorch] Outline OperatorEntry::assertSignatureIsCorrect fail path
* #51247 [PyTorch] check isValidUnboxed() in the dispatcher
* #51165 [PyTorch] Pass TensorOptions by value
* #51245 [PyTorch] Add C10_ALWAYS_INLINE to critical dispatcher paths
* **#51163 [PyTorch] Refactor Dispatcher to inline less code in fast path**

The Dispatcher seems to have been in a precarious local
maximum: I tried to make several different changes to parameter
passing and ended up with regressions due to reduced inlining that
swamped any gains I might have gotten from the parameter passing
changes.

This diff reduces the amount of inline code on the fast path. It
should both reduce code size and provide a platform for making further
improvements to the dispatcher code.

It is a slight performance regression, but it unblocked the following
two diffs (which seem to get us back where we were) from landing.

Differential Revision: [D26091377](https://our.internmc.facebook.com/intern/diff/D26091377/)

swolchok committed:
[PyTorch] Refactor Dispatcher to inline less code in fast path

The Dispatcher seems to have been in a precarious local
maximum: I tried to make several different changes to parameter
passing and ended up with regressions due to reduced inlining that
swamped any gains I might have gotten from the parameter passing
changes.

This diff reduces the amount of inline code on the fast path. It
should both reduce code size and provide a platform for making further
improvements to the dispatcher code.

It is a slight performance regression, but it unblocked the following
two diffs (which seem to get us back where we were) from landing.

Differential Revision: [D26091377](https://our.internmc.facebook.com/intern/diff/D26091377/)

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -384,6 +392,15 @@ inline Return Dispatcher::callWithDispatchKey(const TypedOperatorHandle<Return(A
+}
+
+template<class Return, class... Args>
+inline Return Dispatcher::callWithDispatchKeySlowPath(const TypedOperatorHandle<Return(Args...)>& op, bool pre_sampled, DispatchKey dispatchKey, const KernelFunction& kernel, Args... args) {
```
Just curious- if everything here is technically slow path, is there a reason not to dump this whole function in `Dispatcher.cpp` to prevent inlining? It looks like you did that, but only for `runRecordFunction()`, which is probably "slower path".

My understanding of the relative levels of hotness in the Dispatcher path:
* hot: `shouldRunRecordFunction`, `kernel.call<...>()`
* less hot: create `RecordFunction` guard and give it inputs
* least hot: `RecordFunction.before`

So does putting just the last of those 3 in a `.cpp` file yield better results than the bottom 2?

swolchok replied:
>  is there a reason not to dump this whole function in Dispatcher.cpp to prevent inlining?

I would be happy in theory to dump it in Dispatcher.cpp, but unfortunately it has to be templated on the type signature of the operator (just like `Dispatcher::call` and `Dispatcher::callWithDispatcherKey`) so that we can call out to `impl::boxArgs`, so it must be visible inline in order for a copy to be instantiated for every distinct operator signature.

It's possible that we could extract `impl::boxArgs(args...)` into a lambda and pass it into a hypothetical `callWithDispatchKeySlowPathHelper` that wasn't templated, but I'm not sure how to do that without introducing a round of copying or moving on the arguments and a `std::function`, which would penalize the profiling use case to uncertain benefit. Maybe we'll need to do it in the future, but I think that the space of inlineable dispatch-related code is bigger than just the Dispatcher and we may be able to see bigger gains for less work in other places.

swolchok commented:
By the way, I've diffed the annotated assembly for `at::empty` (which is where `Dispatcher::call{,WithDispatchKey}` end up getting inlined) in a benchmark that calls `torch::empty` in a loop, and it looks like the source of the regression is that we end up with slightly more register pressure for whatever reason (haven't checked exactly why but I suspect we are leaving one or more things in a register in case we have to call into `callWithDispatchKeySlowPath` even though it's behind a C10_UNLIKELY branch; register allocation is [hard](https://en.wikipedia.org/wiki/Register_allocation#Common_problems_raised_in_register_allocation) and we can't expect it to be perfect) and have to reload the address of the relevant `OperatorEntry` in between checking `KernelFunction::isValid` (checks if the `boxed_kernel_func_ != nullptr`) and executing `KernelFunction::call` (loads up `unboxed_kernel_func_` instead and checks it against nullptr).

I will see if I can claw this regression back in a separate PR.

swolchok committed:
Update on "[PyTorch] Refactor Dispatcher to inline less code in fast path"


The Dispatcher seems to have been in a precarious local
maximum: I tried to make several different changes to parameter
passing and ended up with regressions due to reduced inlining that
swamped any gains I might have gotten from the parameter passing
changes.

This diff reduces the amount of inline code on the fast path. It
should both reduce code size and provide a platform for making further
improvements to the dispatcher code.

It is a slight performance regression, but it unblocked the following
two diffs (which seem to get us back where we were) from landing.

Differential Revision: [D26091377](https://our.internmc.facebook.com/intern/diff/D26091377/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Refactor Dispatcher to inline less code in fast path"


The Dispatcher seems to have been in a precarious local
maximum: I tried to make several different changes to parameter
passing and ended up with regressions due to reduced inlining that
swamped any gains I might have gotten from the parameter passing
changes.

This diff reduces the amount of inline code on the fast path. It
should both reduce code size and provide a platform for making further
improvements to the dispatcher code.

It is a slight performance regression, but it unblocked the following
two diffs (which seem to get us back where we were) from landing.

Differential Revision: [D26091377](https://our.internmc.facebook.com/intern/diff/D26091377/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Refactor Dispatcher to inline less code in fast path"


The Dispatcher seems to have been in a precarious local
maximum: I tried to make several different changes to parameter
passing and ended up with regressions due to reduced inlining that
swamped any gains I might have gotten from the parameter passing
changes.

This diff reduces the amount of inline code on the fast path. It
should both reduce code size and provide a platform for making further
improvements to the dispatcher code.

It is a slight performance regression, but it unblocked the following
two diffs (which seem to get us back where we were) from landing.

Differential Revision: [D26091377](https://our.internmc.facebook.com/intern/diff/D26091377/)

[ghstack-poisoned]
